### ISSUE TITLE
Form Data not documented properly using RestAssured

### ISSUE THREAD

**Issue by CalamarBicefalo** [ePages-de/restdocs-api-spec#145]
Support for proper documentation of `application/x-www-form-urlencoded` was added in #137 in response to #113.

However, the solution implemented does not work with specs generated with RestAssured.

The reason for this is restassured expects you to use formParam in order to specify `application/x-www-form-urlencoded`. 
- If done as documented, then the body is treated as query params in the openapi spec which is incorrect
- If a body is used instead - to force the solution implemented in #137, Spring Rest docs cannot document those properties (the test fails). 
- If a 'hacky' body is added in addition to `formParam` - to force the solution implemented in #137 , restassured fails pointing out either body or formParam can be used, but not both.

Long story short, there's no way to create a valid openapi spec with a `application/x-www-form-urlencoded` body using RestAssured.

To put words in code:

Given the current test - the legit way of doing this according to spring docs + a content type to make #137  happy

```kotlin
@Test
    fun `post-token-client-credentials`() {
        RestAssured.given(this.spec).urlEncodingEnabled(true)
                // this is how x-www-form-urlencoded params are added in RestAssured according to the spring docs
                .formParam("grant_type", "client_credentials")
                .formParam("client_id", clientId)
                .formParam("client_secret", clientSecret)
                .contentType(ContentType.URLENC)
                .filter(
                        document(
                                identifier = "{method-name}",
                                //... stuff irrelevant for this issue
                                snippets = *arrayOf(tokenResponseFields,
                                        requestParameters(
                                               // these are the three bad boys that 
                                               // need to be documented this way according to spring
                                                grantTypeParam,
                                                clientIdParam,
                                                clientSecretParam
                                        ))
                        )
                )
                .`when`().post("/v1/token").apply { prettyPrint() }
                .then().assertThat().statusCode(CoreMatchers.`is`(200))
    }
```

The whole thing falls apart when building the data model because of this line:

com/epages/restdocs/apispec/ResourceSnippet.kt:68
```kotlin
if (hasRequestBody) getContentTypeOrDefault(operation.request.headers) else null
```

Unfortunately `hasRequestBody` says `false`. If you try to set the actual body as body (instead of `formParams`) spring fails.

Any ideas?

_Originally posted by @CalamarBicefalo in https://github.com/ePages-de/restdocs-api-spec/issues/113#issuecomment-706205489_

**Comment by ozscheyge** [comment]
Seems like a deadlock between RestAssured, Spring restdocs and OpenAPI, thus I don't see, how restdocs-api-spec should behave in this case, either. :confused:

**Comment by CalamarBicefalo** [comment]
Hey @ozscheyge, although I agree with what you said, to me we reached this "deadlock" because perhaps we are missing the right abstractions when we put in place the PR to support this feature.

The fact that we reuse path variable objects to create this spec to me is a smell, and instead, I think we should have an appropriate abstraction that models form-urlencoded payloads. I think that's the case because that IS part of the openapi spec domain, and therefore, should be modelled by this library. It is neither Spring nor RestAssured responsibility to do so.

So basically, in order to resolve this, I'd lean towards:

1. Undoing the "workarounds" of the PR
2. Creating an abstraction for form url encoded payloads
3. Creating adapters for the various clients with their peculiarities resolved there

It is a bit of heavy lifting, but it seems like the right thing to me. 

Also, as an unrelated note, I find hard to determine how to get some e2e tests for the entire flow. Do you have some of those? If so where are they?

**Comment by ozscheyge** [comment]
I'm afraid, "missing the right abstractions" is by design of restdocs-api-spec.
The basic approach is to dump all information of common Spring restdocs descriptors into a single JSON snippet (`resource.json`) for further transformation to the target formats via gradle-plugins (as opposed to generating multiple adoc/markdown snippets and running asciidoctor).
This is achieved by being API-compatible with the Spring restdocs `document` functions. Thus replacing an import is sufficient: all existing descriptors are used as-is, tests and documentation in tests don't need to be adapted.

Afaics (research done by @afronski  in https://github.com/ePages-de/restdocs-api-spec/issues/113 ), the Spring restdocs `ParameterDescriptor` should be used to document form-urlencoded payloads. The respective gradle-plugin is then responsible for inferring missing information ("is it a path, query or request body parameter? Do I need to infer a schema? For which part of the request response is this schema?") which is needed for the target format.

I see no way to improve the abstraction done by restdocs-api-spec which is constrained by the common Spring restdocs request/response snippets and descriptors (if we'd lack support for some descriptors or descriptor properties -- that could be fixed!)
To improve the abstraction or raw format, one could introduce another new snippet or preprocessor adding/requiring the needed information. However, that might as well be a separate project, since it would not be swap-import-compatible with Spring restdocs.

**Comment by ozscheyge** [comment]
Regarding e2e tests:
I assume, you mean an automatic test generating a target spec format from a MockMvc/RestAssured test and then successfully importing this to some tool or validator?
Indeed, we do not have that. 

We use restdocs-api-spec (not necessarily latest version) in production to generate a Postman Collection for our API:
https://developer.epages.com/beyond-docs/#postman_collection
Direct dl link: http://api-docs.beyondshop.cloud/beyond-rest-api-public.json
...and we diff the OpenAPI 3 specs in regular intervals as a starting point for our semi-automatic changelog: https://developer.epages.com/beyond-docs/#change_log
In the past, we used https://stoplight.io/ SAAS solution with OpenAPI 2. It was fully automated, so a successful build and deployment would also update stoplight.io API docs and we would have noticed if the format wasn't valid. Ultimately, we dialed back from that one due to OpenAPI 2 not being sufficient for us (and stoplight not having proper OpenAPI 3 support at that time).

Then again, I see no big value in having a full e2e test. What current tests ensure:
* The document function called with Spring restdocs descriptors yields the desired restdocs-api-spec model and snippet (`resource.json`).
* Based on the restdocs-api-spec model the target format is generated as desired by the respective gradle-plugins.

We don't automatically assert, whether some tool can process the formats we produce, but we can assert anything about the formats we produce. So if there's an error, it should be possible to write a test, fixing it and hardening the project.

**Comment by CalamarBicefalo** [comment]
Ok, I may try to see what can be done.

regarding resource.json. Is there a test (somewhere) where I can start with a mockmvc/restassured spec and end up with a `resource.json` file? Just would like to prove the issue in your test suite before changing anything really...

**Comment by ozscheyge** [comment]
Sure, that'd be great @CalamarBicefalo !


* Tests for the MockMvc/RestAssured `document` wrapper methods, just testing that `resource.json` file is generated:
  https://github.com/ePages-de/restdocs-api-spec/blob/master/restdocs-api-spec-mockmvc/src/test/kotlin/com/epages/restdocs/apispec/MockMvcRestDocumentationWrapperIntegrationTest.kt
  https://github.com/ePages-de/restdocs-api-spec/blob/master/restdocs-api-spec-restassured/src/test/kotlin/com/epages/restdocs/apispec/RestAssuredRestDocumentationWrapperIntegrationTest.kt
* Tests asserting that `resource.json` is generated according to restdocs-api-spec's model:
  https://github.com/ePages-de/restdocs-api-spec/blob/master/restdocs-api-spec/src/test/kotlin/com/epages/restdocs/apispec/ResourceSnippetTest.kt
* The model itself isn't further tested, but can be found here:
  https://github.com/ePages-de/restdocs-api-spec/blob/master/restdocs-api-spec-model/src/main/kotlin/com/epages/restdocs/apispec/model/ResourceModel.kt

**Comment by ozscheyge** [comment]
Closing this due to inactivity.

Problem and reason are explained, contribution path is outlined.

Also:

> Please note that this extension was developed for JSON-based APIs. We do not expect this extension to build usable API specification for non-JSON request or response bodies.

https://github.com/ePages-de/restdocs-api-spec/blob/master/README.md
